### ISSUE TITLE
[Bug] Noticing some anomalies in the training eval with YourTTS training

### ISSUE THREAD

**Issue by neural-loop** [coqui-ai/TTS#2569]
### Describe the bug

The tensorboard shows that the first two or last two columns in the EvalFigures/evalspectrogram/real seem to be "multiplied" in some instances. This does not happen every time, but when it does, it shows up in the 'diff'. This may potentially affect the training output and cause artifacts.

The images show that the multiplication occurs uniformly across the entire spectral band. I am investigating this issue and trying to determine its impact. I estimate that the affected area is around 6% (two out of 32 columns) or 12% (four out of 32 columns). I am not sure how many frames are impacted as the tensorboard only shows a snapshot of the particular step. However, this issue occurs frequently throughout different datasets.

![image](https://user-images.githubusercontent.com/654993/235451893-fde93672-4863-4871-9762-dd439c6359a6.png)
![image](https://user-images.githubusercontent.com/654993/235452366-f08cdf9d-3c0b-4c73-930a-55c2b5608b42.png)
![image](https://user-images.githubusercontent.com/654993/235474518-ce37d559-0f9f-40fc-b4fa-464d02a86122.png)

Maybe it is related to the beginning or end of a file, or something along those lines and some padding has to be inserted to complete the segment? Am curious to know what it is :) I am using mixed precision, I'll test later without to see if that makes any difference.

### To Reproduce

Training on a YourTTS (and possibly VITS) model

### Expected behavior

I would expect that in the 'real' data and 'diff', that there wouldn't be an section of the spectrogram that intermittently has some consistent difference - as that would suggest that it's not something from the training data itself. 

### Environment

```shell
env_info.py
{
    "CUDA": {
        "GPU": [
            "NVIDIA GeForce RTX 2080 Ti"
        ],
        "available": true,
        "version": "11.7"
    },
    "Packages": {
        "PyTorch_debug": false,
        "PyTorch_version": "1.13.1+cu117",
        "TTS": "0.13.0",
        "numpy": "1.23.5"
    },
    "System": {
        "OS": "Linux",
        "architecture": [
            "64bit",
            "ELF"
        ],
        "processor": "x86_64",
        "python": "3.10.6",
        "version": "#42~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Apr 18 17:40:00 UTC 2"
    }
}
```

**Comment by neural-loop** [comment]
I outputted some of the mels from vits wav_to_mel. While some times there is a border, I did not see anything that stood out as much as the two columns 

![0000_2__1024_80_16000_0_None](https://github.com/coqui-ai/TTS/assets/654993/1862f78d-4fdd-4823-a4c2-03568e9ea9ac)
![0001_2__1024_80_16000_0_None](https://github.com/coqui-ai/TTS/assets/654993/684b4f18-38b4-4901-ac92-8b06e42a13d3)

![0002_2__1024_80_16000_0_None](https://github.com/coqui-ai/TTS/assets/654993/4dbf23fe-9fcb-4902-ac2e-828384fc393a)
![0003_2__1024_80_16000_0_None](https://github.com/coqui-ai/TTS/assets/654993/45a3fec9-06c9-4612-8a83-b6866f7d4730)

![0004_2__1024_80_16000_0_None](https://github.com/coqui-ai/TTS/assets/654993/d156378e-cd00-43bf-9c36-ccf8f996dd48)

Still not sure if there actually is an issue, but I will work my way through to the output found on the tensorboard. I noticed there are some properties such as reflect, and padding, and shifting around so maybe something could be introduced somewhere.

**Comment by fkwlqm** [comment]
can you explain what do you mean by the first two columns being multiplied? Thanks

**Comment by neural-loop** [comment]
Multiply probably isn't technically correct, though what I was referring to is the blending mode effect I've seen in photoshop which multiplies the luminosity of the base color by the blend color. 

![image](https://github.com/coqui-ai/TTS/assets/654993/2ed61728-2591-4eea-9e81-c187b080b3d8)

Looking down the leftmost two columns, I can't tell 100% but it doesn't seem they are brighter uniformly, so I was looking in the code where reflect is used, thinking that there possibly a calculation that results in a decimal somewhere, or where padding is applied that  produces some overlap that is reflected back on top of the base spectrogram, and that results in a couple columns appearing lighter. 

So the thread I was exploring is where there should be 32 columns, 36 are produced, and the extra 2 on each side wrap back on top of 1 and 2 and 31 and 32.

**Comment by erogol** [comment]
This is interesting. I'll try to debug if I find some time. It might be about the settings used for the mel computation. I agree that it might introduce artifacts.

**Comment by erogol** [comment]
@WeberJulian @Edresson any idea?

**Comment by neural-loop** [comment]
I wasn't able to figure out a way to produce it consistently. It helped that the top rows of the spectrogram were darker (maybe because my audio was at -25db compared to the train data at -27) so that the difference is visible in those top corners. It's imperceptible when the spectrogram fills the frame. It also doesn't appear on all of the frames, just enough of them that as I scroll through tensorboard steps I notice it frequently.

**Comment by erogol** [comment]
@WeberJulian @Edresson any idea?

**Comment by Edresson** [comment]
> @WeberJulian @Edresson any idea?

I think that it is possible this issue be related to the padding that we do [here](https://github.com/coqui-ai/TTS/blob/a4860eac9a8fcb5240de2fd43d8bad62f1d3d10b/TTS/tts/models/vits.py#L116).

**Comment by neural-loop** [comment]
It might be related to mixed_precision

**Comment by phamkhactu** [comment]
Hi @neural-loop,

Have you found any information?? I've trained  vits. After training 300 epochs, I find that model swallow phoneme and faster speed?

**Comment by neural-loop** [comment]
@phamkhactu I've experimented with a couple solution attempts but haven't solved it yet. 

One attempt was with ChatGPT (didn't seem to solve, thinking about trying similar with the mel functions that also use reflect):

> It seems that the issue you're observing might be related to the use of padding with "reflect" mode in the wav_to_spec function. The reflection padding is often used to handle edge effects when calculating the short-time Fourier transform (STFT) spectrogram, but it can sometimes introduce artifacts at the boundaries.
> 
> To avoid these artifacts, you can try using "constant" padding with a small constant value (e.g., zero) instead of "reflect". This way, the padding won't affect the STFT calculation and should not introduce any unwanted artifacts. You can modify the wav_to_spec function to use "constant" padding as follows:

```
def wav_to_spec(y, n_fft, hop_length, win_length, center=False):
    """
    Args Shapes:
        - y : :math:`[B, 1, T]`

    Return Shapes:
        - spec : :math:`[B,C,T]`
    """
    y = y.squeeze(1)

    if torch.min(y) < -1.0:
        print("min value is ", torch.min(y))
    if torch.max(y) > 1.0:
        print("max value is ", torch.max(y))

    global hann_window
    dtype_device = str(y.dtype) + "_" + str(y.device)
    wnsize_dtype_device = str(win_length) + "_" + dtype_device
    if wnsize_dtype_device not in hann_window:
        hann_window[wnsize_dtype_device] = torch.hann_window(win_length).to(dtype=y.dtype, device=y.device)

    pad_size = int((n_fft - hop_length) / 2)
    y = torch.nn.functional.pad(y.unsqueeze(1), (pad_size, pad_size), mode="constant", value=0)
    y = y.squeeze(1)

    spec = torch.stft(
        y,
        n_fft,
        hop_length=hop_length,
        win_length=win_length,
        window=hann_window[wnsize_dtype_device],
        center=center,
        pad_mode="constant",
        normalized=False,
        onesided=True,
        return_complex=False,
    )

    spec = torch.sqrt(spec.pow(2).sum(-1) + 1e-6)
    return spec

```

**Comment by phamkhactu** [comment]
@neural-loop, I will try out it, if I have any result, I will share.

@erogol @Edresson 
And now, I have try out glow-tts , vits. I have some pros and cons:
- Glow: speak is very natural but buzzing noise 
- Vits: not buzzing noise however  speak is not natural. 
Do you try on some models? and you have any suggestion?

**Comment by neural-loop** [comment]
You may find the answers you are seeking on the Coqui discord, it's outside
the scope of this issue

**Comment by erogol** [comment]
@phamkhactu you might need to train a vocoder for glow tts.
